util/DynamicFunction.hpp

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+#pragma once
+
+#include <libloaderapi.h>
+#include <wil/resource.h>
+#include <functional>
+
+template <class T>
+class DynamicFunction;
+
+/// @brief Wrapper for a runtime dynamically loaded function
+template <class R, class... Args>
+class DynamicFunction<R(Args...)>
+{
+public:
+    DynamicFunction(const wil::shared_hmodule& module, const std::string& functionName)
+        : m_module{module}, m_function{LoadFunction(module, functionName)}
+    {
+    }
+
+    DynamicFunction(const std::wstring& moduleName, const std::string& functionName)
+        : DynamicFunction{LoadModule(moduleName), functionName}
+    {
+    }
+
+    decltype(auto) operator()(Args... args) const
+    {
+        return m_function(std::forward<Args>(args)...);
+    }
+
+private:
+    static wil::shared_hmodule LoadModule(const std::wstring& name)
+    {
+        wil::shared_hmodule module{LoadLibraryEx(name.c_str(), nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32)};
+        THROW_LAST_ERROR_IF(!module);
+        return module;
+    }
+
+    static std::function<R(Args...)> LoadFunction(const wil::shared_hmodule& module, const std::string& name)
+    {
+        std::function<R(Args...)> function = reinterpret_cast<R (*)(Args...)>(GetProcAddress(module.get(), name.c_str()));
+        THROW_LAST_ERROR_IF(!function);
+        return function;
+    }
+
+private:
+    std::function<R(Args...)> m_function;
+    wil::shared_hmodule m_module;
```

Comment:
This is shared because another way to use this class is to load the library once first and reuse it to load multiple functions (first constructor). I am not using it this way (the module is only actually loaded once even with multiple calls to `LoadLibraryEx`), but I kept the two options from WSL.

If you think it a unique_hmodule would be better, I can use it instead.

Reply:
IMHO, it's preferrable to use the type that express intent - shared_* expresses the intent that this variable (and object) can be shared across multiple instances. If that's not deliberately needed, I start with unique_*

Reply:
Ok. Since it's not needed for now, I'll move to unique and remove the unused constructor.